aslprep/utils/cbf.py

```diff
@@ -819,3 +819,335 @@ def estimate_t1(metadata):
+    plds
+    tau : :obj:`float` or :obj:`numpy.ndarray`
+        Label duration. May be a single value or may vary across volumes/PLDs.
+    t1blood
+    labeleff : :obj:`float`
+        Labeling efficiency, also referred to as alpha.
+        In the case of background suppression, this combines alpha and alpha_bs,
+        so we don't need a separate term for alpha_bs.
+    scaled_m0data
+    partition_coefficient
+
+    Returns
+    -------
+    cbf
+    att
+    abat
+    abv
+
+    Notes
+    -----
+    This model does not include a separate term for T1,tissue, but we could expand it to do so
+    in the future.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    import scipy
+
+    n_voxels, n_volumes = deltam_arr.shape
+    if deltam_arr.shape != plds.shape:
+        raise ValueError(
+            f"Number of PostLabelingDelays ({plds.shape}) does not match "
+            f"number of delta-M volumes ({deltam_arr.shape})."
+        )
+
+    if scaled_m0data.size != n_voxels:
+        raise ValueError(
+            f"Number of voxels in M0 data ({scaled_m0data.size}) does not match "
+            f"number of delta-M voxels ({n_voxels})."
+        )
+
+    if is_casl:
+        assert ti1 is None
+        assert tau is not None
+        if isinstance(tau, float):
+            tau = np.full((n_volumes,), tau)
+        else:
+            tau = np.ndarray(tau)
+
+            if tau.size != n_volumes:
+                raise ValueError(
+                    f"Number of values in tau ({tau.shape}) does not match number of "
+                    f"delta-M volumes ({n_volumes})."
+                )
+    else:
+        assert ti1 is not None
+        assert tau is None
+
+    m0_a = scaled_m0data / partition_coefficient
+
+    cbf = np.zeros(n_voxels)
+    att = np.zeros(n_voxels)
+    abat = np.zeros(n_voxels)
+    abv = np.zeros(n_voxels)
+    for i_voxel in range(n_voxels):
+        deltam_voxel = deltam_arr[i_voxel, :]
+        plds_voxel = plds[i_voxel, :]
+        m0_a_voxel = m0_a[i_voxel]
+
+        # TODO: Figure out what alpha_bs and m0_b should be.
+        # The independent variables used to estimate cbf, etc. are either floats or arrays,
+        # but curve_fit needs them all to be the same size/shape.
+        xdata = np.zeros((n_volumes, 6))
+        xdata[0, 0] = labeleff
+        xdata[0, 1] = t1blood
+        xdata[0, 2] = m0_a_voxel
+        xdata[0, 3] = m0_a_voxel
+        xdata[:, 4] = plds_voxel
+
+        if is_casl:
+            xdata[:, 5] = tau
+            func = calculate_deltam_pcasl
+        else:
+            xdata[0, 5] = ti1
+            func = calculate_deltam_pasl
+
+        popt = scipy.optimize.curve_fit(
+            func,
+            xdata=xdata,
+            ydata=deltam_voxel,
+            # initial estimates for DVs (cbf, att, abat, abv)
+            p0=(1, 1, 1, 1),
+            # lower and upper bounds for DVs
+            bounds=((0, 0, 0, 0), (np.inf, np.inf, np.inf, np.inf)),
```

Comment:
Need to come up with better upper bounds. aBAV is probably on a 0 - 1 or 0 - 100 scale, right?